scrapy/downloadermiddlewares/httpauth.py

```diff
@@ -15,17 +16,28 @@ class HttpAuthMiddleware:
 
     @classmethod
     def from_crawler(cls, crawler):
-        o = cls()
+        o = cls()   
         crawler.signals.connect(o.spider_opened, signal=signals.spider_opened)
         return o
 
     def spider_opened(self, spider):
-        usr = getattr(spider, 'http_user', '')
-        pwd = getattr(spider, 'http_pass', '')
-        if usr or pwd:
-            self.auth = basic_auth_header(usr, pwd)
+        pass
+
+    def _authenticated(self, response):
+        if response.status == "401":
+            return False
+
+        return True
 
     def process_request(self, request, spider):
-        auth = getattr(self, 'auth', None)
-        if auth and b'Authorization' not in request.headers:
+        if _authenticated(self, Response(request.url)):
```

Comment:
This condition will always be true, because the response is created here without a status, hence its status will be 200 by default.

Reply:
Should I check for response or only just request headers? I believe middleware should retry authentication 1 or 2 times if the authorization fails.